lib/lower/ASTBuilder.cpp

```diff
@@ -0,0 +1,783 @@
+              return AST::Evaluation{x.value(), parents.back(), s.source, {}};
+            },
+        },
+        s.statement.u);
+  }
+
+  template <typename A>
+  AST::Evaluation makeEvalIndirect(const Pa::Statement<Co::Indirection<A>> &s) {
+    return AST::Evaluation{s.statement.value(), parents.back(), s.source,
+                           s.label};
+  }
+
+  template <typename A>
+  AST::Evaluation makeEvalDirect(const Pa::Statement<A> &s) {
+    return AST::Evaluation{s.statement, parents.back(), s.source, s.label};
+  }
+
+  // When we enter a function-like structure, we want to build a new unit and
+  // set the builder's cursors to point to it.
+  template <typename A>
+  bool enterFunc(const A &f) {
+    auto &unit = addFunc(AST::FunctionLikeUnit{f, parents.back()});
+    funclist = &unit.funcs;
+    pushEval(&unit.evals);
+    parents.emplace_back(&unit);
+    return true;
+  }
+
+  void exitFunc() {
+    popEval();
+    funclist = nullptr;
+    parents.pop_back();
+  }
+
+  // When we enter a construct structure, we want to build a new construct and
+  // set the builder's evaluation cursor to point to it.
+  template <typename A>
+  bool enterConstruct(const A &c) {
+    auto &con = addEval(AST::Evaluation{c, parents.back()});
+    con.subs = new std::list<AST::Evaluation>();
+    pushEval(con.subs);
+    parents.emplace_back(&con);
+    return true;
+  }
+
+  void exitConstruct() {
+    popEval();
+    parents.pop_back();
+  }
+
+  // When we enter a module structure, we want to build a new module and
+  // set the builder's function cursor to point to it.
+  template <typename A>
+  bool enterModule(const A &f) {
+    auto &unit = addUnit(AST::ModuleLikeUnit{f, parents.back()});
+    funclist = &unit.funcs;
+    parents.emplace_back(&unit);
+    return true;
+  }
+
+  void exitModule() {
+    funclist = nullptr;
+    parents.pop_back();
+  }
+
+  template <typename A>
+  A &addUnit(const A &unit) {
+    pgm->getUnits().emplace_back(unit);
+    return std::get<A>(pgm->getUnits().back());
+  }
+
+  template <typename A>
+  A &addFunc(const A &func) {
+    if (funclist) {
+      funclist->emplace_back(func);
+      return funclist->back();
+    }
+    return addUnit(func);
+  }
+
+  /// move the Evaluation to the end of the current list
+  AST::Evaluation &addEval(AST::Evaluation &&eval) {
+    assert(funclist && "not in a function");
+    assert(evallist.size() > 0);
+    evallist.back()->emplace_back(std::move(eval));
+    return evallist.back()->back();
+  }
+
+  /// push a new list on the stack of Evaluation lists
+  void pushEval(std::list<AST::Evaluation> *eval) {
+    assert(funclist && "not in a function");
+    assert(eval && eval->empty() && "evaluation list isn't correct");
+    evallist.emplace_back(eval);
+  }
+
+  /// pop the current list and return to the last Evaluation list
+  void popEval() {
+    assert(funclist && "not in a function");
+    evallist.pop_back();
+  }
+
+  AST::Program *pgm;
```

Comment:
Either a `unique_ptr` or reference would be safer here.

Reply:
That was updated while you commented.